pkg/action/release_testing.go

```diff
@@ -45,39 +40,21 @@ func NewReleaseTesting(cfg *Configuration) *ReleaseTesting {
 }
 
 // Run executes 'helm test' against the given release.
-func (r *ReleaseTesting) Run(name string) error {
+func (r *ReleaseTesting) Run(name string) (*release.Release, error) {
 	if err := validateReleaseName(name); err != nil {
-		return errors.Errorf("releaseTest: Release name is invalid: %s", name)
+		return nil, errors.Errorf("releaseTest: Release name is invalid: %s", name)
 	}
 
 	// finds the non-deleted release with the given name
 	rel, err := r.cfg.Releases.Last(name)
 	if err != nil {
-		return err
+		return rel, err
 	}
 
 	if err := r.cfg.execHook(rel, release.HookTest, r.Timeout); err != nil {
 		r.cfg.Releases.Update(rel)
-		return err
-	}
-
-	if r.Cleanup {
```

Comment:
First off, thank you for taking the time to refactor `helm test`. We really appreciate the work you've put into this.

That being said, we cannot remove this feature and replace it with hook annotations. This feature serves a very different purpose than what the hook annotations provide, which is why `helm test` was not originally designed around the hook system.

A very simple test where this flag is required:

1. I have a chart that someone else has built (e.g. `stable/wordpress`)
1. I want to run the tests
1. I want to have the test hooks cleaned up in preparation for the next test run

Prior to this change, the use case was done using the following commands:

```
$ helm install wordpress stable/wordpress
$ helm test run wordpress --cleanup
```

Now with this functionality removed, a user must now perform the following:

```
$ helm fetch stable/wordpress --untar
$ $EDITOR wordpress/templates/tests/
$ helm install wordpress ./wordpress
$ helm test wordpress
```

So now, the user must MODIFY the chart before installing to handle the use case. If they don't want to clean up test hooks after the test run, then they have to modify the chart AGAIN, run `helm upgrade` and finally run `helm test`. This is a massive barrier of entry for users who just want their test hooks to be cleaned up after running `helm test`. It's for that reason this functionality must remain as-is. I don't see a way around this with the way hook annotations work today, sadly.

Reply:
Perhaps we should schedule some time to discuss this; I think the workflow concerns here are addressed by https://github.com/helm/helm/issues/6365.

When those changes are combined with https://github.com/helm/helm/issues/6364 and the proposal to amend the default `hook-delete-policy` from https://github.com/helm/helm/issues/6020 I think we should be able to deliver a much nicer experience than Helm 2 out of the box. This PR is not representative of the whole picture.

Reply:
Discussed offline. Workflow concerns are indeed addressed through other proposed paths.